dot/parachain/candidate-validation/candidate_validation.go

```diff
@@ -1,27 +1,99 @@
-// Copyright 2023 ChainSafe Systems (ON)
+// Copyright 2024 ChainSafe Systems (ON)
 // SPDX-License-Identifier: LGPL-3.0-only
 
-package parachain
+package candidatevalidation
 
 import (
 	"bytes"
+	"context"
 	"errors"
 	"fmt"
+	"sync"
 
 	parachainruntime "github.com/ChainSafe/gossamer/dot/parachain/runtime"
 	parachaintypes "github.com/ChainSafe/gossamer/dot/parachain/types"
+	"github.com/ChainSafe/gossamer/internal/log"
 	"github.com/ChainSafe/gossamer/lib/common"
-
 	"github.com/ChainSafe/gossamer/pkg/scale"
 )
 
+var logger = log.NewFromGlobal(log.AddContext("pkg", "parachain-candidate-validation"))
+
 var (
 	ErrValidationCodeMismatch   = errors.New("validation code hash does not match")
 	ErrValidationInputOverLimit = errors.New("validation input is over the limit")
 )
 
+type CandidateValidation struct {
+	wg       sync.WaitGroup
+	stopChan chan struct{}
+
+	SubsystemToOverseer chan<- any
+	OverseerToSubsystem <-chan any
+}
+
+func (cv *CandidateValidation) Run(ctx context.Context, OverseerToSubSystem chan any, SubSystemToOverseer chan any) {
+	cv.wg.Add(1)
+	go cv.processMessages(&cv.wg)
+}
+
+func (*CandidateValidation) Name() parachaintypes.SubSystemName {
+	return parachaintypes.CandidateValidation
+}
+
+func (*CandidateValidation) ProcessActiveLeavesUpdateSignal(signal parachaintypes.ActiveLeavesUpdateSignal) error {
+	// NOTE: this subsystem does not process active leaves update signal
+	return nil
+}
+
+func (*CandidateValidation) ProcessBlockFinalizedSignal(signal parachaintypes.BlockFinalizedSignal) error {
+	// NOTE: this subsystem does not process block finalized signal
+	return nil
+}
+
+func (cv *CandidateValidation) Stop() {
+	close(cv.stopChan)
+	cv.wg.Wait()
+}
+
+func (cv *CandidateValidation) processMessages(wg *sync.WaitGroup) {
+	defer wg.Done()
+	for {
+		select {
+		case msg := <-cv.OverseerToSubsystem:
+			logger.Debugf("received message %v", msg)
+			switch msg := msg.(type) {
+			case ValidateFromChainState:
+				// TODO: implement functionality to handle ValidateFromChainState, see issue #3919
+			case ValidateFromExhaustive:
+				// TODO: implement functionality to handle ValidateFromExhaustive, see issue #3547
+			case PreCheck:
+				// TODO: implement functionality to handle PreCheck, see issue #3921
+
+			case parachaintypes.ActiveLeavesUpdateSignal:
+				err := cv.ProcessActiveLeavesUpdateSignal(msg)
+				if err != nil {
+					logger.Errorf("failed to process active leaves update signal: %w", err)
+				}
```

Comment:
this function returns nil. just ignore the returned value.
```suggestion
				_ = cv.ProcessActiveLeavesUpdateSignal(msg)
```

Reply:
Updated.